docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,338 @@
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. **Provide a new "ServerSideDefaultsConfig" struct for AWS service teams to provide which spec fields(spec.A) or spec field members(spec.A.X) will be late initialized.**
+
+    Only the fields/members mentioned as part of 'ServerSideDefaultsConfig' will be considered for late initialization. (The reason is explained inside step 3 "NOTE" )
+
+    `ServerSideDefaultsConfig` struct will be the member of `ResourceConfig` and it will look like following:
+
+    ```
+    type ServerSideDefaultsConfig struct { 
+    
+    // valid values: 'readOne, readMany, create, update'
+    DefaultSourceMethod *string `json:"default_source_method"`
+    
+    DefaultedSpecFields []DefaultedSpecFieldConfig `json:"defaulted_spec_fields"`
+    }
+
+
+    type DefaultedSpecFieldConfig struct {
+    
+    // valid values: 'readOne, readMany, create, update'
+    // If this is not mentioned, then default_source_method is used
+    SourceMethod *string `json:"source_method,omitempty"`
+    
+    // Path to the field relative to Spec, which will be late initialized
+    // Example for primitives: "a" to replace "desired.ko.Spec.a" with "<source_method>Latest.ko.spec.a"
+    // Example for struct: "a.b" to replace "desired.ko.Spec.a.b" with "<op_name>Latest.ko.spec.a.b"
+    // Example for map: "a.b.[c]" to replace "desired.ko.Spec.a.b[c]" with "<op_name>Latest.ko.spec.a.b[c]"
+    // TODO: support for List
+    
+    Path *string `json:"path"` 
+    
+    // TODO: possible support for source path and target path, which do not have to be exactly similar as they are at present.
+    }
+    ```
+
+2. **Provide 3 new hooks to AWS service teams to customize the code generation in step 5**
+    
+    These 3 hooks will be:
+    
+    a) **server_side_defaults_initialization_override**: This hook will provide the **complete implementation** for setting default values of optional spec fields which are not provided by the k8s user and defaulted by AWS service.
+    This hook is mutually exclusive with "pre_server_side_defaults_initialization" & "post_server_side_defaults_initialization" hooks.
+
+    b) **pre_server_side_defaults_initialization**:- This hook will provide the custom code to be inserted **before** setting server side defaults for missing optional spec fields.
+
+    c) **post_server_side_defaults_initialization**: This hook will provide the custom code to be inserted **after** setting server side defaults for missing optional spec fields.
+
+3. **Updating Spec fields in sdkCreate and sdkUpdate methods**
```

Comment:
I guess I don't understand why we want to make any changes in the `sdkCreate` and `sdkUpdate` methods. Instead, what I thought was being added was a change to the `resourceReconcilers.Sync` method that would call `ReadOne`  the resource manager immediately after calling Create when there were fields in the generator config marked as `DefaultImmediately` **AND** when the user had not specified those fields in the desired CR already.

In other words, we would just change the `resourceReconciler.Sync` method implementation from this:

```go
	rlog.Enter("rm.ReadOne")
	latest, err = rm.ReadOne(ctx, desired)
	rlog.Exit("rm.ReadOne", err)
	if err != nil {
		if err != ackerr.NotFound {
			if latest != nil {
				// this indicates, that even though ReadOne failed
				// there is some changes available in the latest.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latest)
			}
			return err
		}
		if isAdopted {
			return ackerr.AdoptedResourceNotFound
		}
		// Before we create the backend AWS service resources, let's first mark
		// the CR as being managed by ACK. Internally, this means adding a
		// finalizer to the CR; a finalizer that is removed once ACK no longer
		// manages the resource OR if the backend AWS service resource is
		// properly deleted.
		if err = r.setResourceManaged(ctx, desired); err != nil {
			return err
		}

		rlog.Enter("rm.Create")
		latest, err = rm.Create(ctx, desired)
		rlog.Exit("rm.Create", err)
		if err != nil {
			if latest != nil {
				// this indicates, that even though Create failed
				// there is some changes available in the latest.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latest)
			}
			return err
		}
		rlog.Info("created new resource")
...
```

to this:

```go
	rlog.Enter("rm.ReadOne")
	latest, err = rm.ReadOne(ctx, desired)
	rlog.Exit("rm.ReadOne", err)
	if err != nil {
		if err != ackerr.NotFound {
			if latest != nil {
				// this indicates, that even though ReadOne failed
				// there is some changes available in the latest.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latest)
			}
			return err
		}
		if isAdopted {
			return ackerr.AdoptedResourceNotFound
		}
		// Before we create the backend AWS service resources, let's first mark
		// the CR as being managed by ACK. Internally, this means adding a
		// finalizer to the CR; a finalizer that is removed once ACK no longer
		// manages the resource OR if the backend AWS service resource is
		// properly deleted.
		if err = r.setResourceManaged(ctx, desired); err != nil {
			return err
		}

		rlog.Enter("rm.Create")
		latest, err = rm.Create(ctx, desired)
		rlog.Exit("rm.Create", err)
		if err != nil {
			if latest != nil {
				// latest will have any fields in the Spec that were set to server-side defaulted
				// values that were in the Create Output shape and any Status fields, so we
				// patch the entire CR here, Spec/Metadata as well as Status.
				_ = r.patchResource(ctx, desired, latest)
			}
			if r.rd.NeedsLateInitialization(latest) {
				// We will get here only if there are any fields in the `latest` CR that
				// are `nil` AND are marked as having an immediately-available
				// server-side default value in the generator config.
				rlog.Enter("rm.ReadOne (LateInitiliazation")
				latestWithDefaults, err = rm.ReadOne(ctx, latest)
				rlog.Exit("rm.ReadOne (LateInitialization)", err)
				if err != nil {
					return err
				}
				// Patch the latestWithDefaults resource back to the Kubernetes API because
				// it will contain the Spec fields set to server-side defaulted values.
				_ = r.patchResource(ctx, latest, latestWithDefaults)
			}
			return err
		}
		rlog.Info("created new resource")
...
```

So, the `AWSResourceDescriptor.NeedsLateInitialization` method is the thing that would be a new interface method and would need to simply return a `bool` if the supplied resource has any fields that are set to `nil` but may have a server-side defaulted value (retrieved via the `AWSResourceManager.ReadOne` call.

Reply:
>"I guess I don't understand why we want to make any changes in the sdkCreate and sdkUpdate methods

My reasoning was that If the late initialized fields are present in the output of create/update response, why make extra readOne calls from reconciler.

And if there are some late initialized fields which will not be part of the Create/Update response, they will be handled separately in AWSResourceManager.SetDefaults() method, which will be called from reconciler.Sync() as mentioned in proposal. 
AWSResourceManager.SetDefaults() will handle those late initialized fields calling ReadOne/Many or any custom code provided by AWS Service teams using hooks.

nit: In the comment, r.patchResource(ctx, latest, latestWithDefaults) , we will have to patch on top of desired correct? so it should be like r.patchResource(ctx, desired, latestWithDefaults)

Reply:
Yeah, I can see your proposed approach working out.

My proposed solution was based on the fact that we do not update Spec fields during AWSResourceManager.Create/Update calls. But Allowing it will simplify handling of server-side-defaults. :+1: